### ISSUE TITLE
[BUG] varea_stack() and vline_stack() fails to update correctly when new source data is different length

### ISSUE THREAD

**Issue by pnetherwood** [bokeh/bokeh#11119]
Bokeh 2.3.0 Python 3.8

I have a vstack_area() chart running in a server. Updating the source data works fine unless the data is is a different length. If the data is shorter the then the xaxis does not update. If the data is longer then no change occurs and the is a size missmatch error in the console. It all works fine if the data is the same length.

Sample code below:

```python
import numpy as np
import pandas as pd
from bokeh.io import curdoc
from bokeh.layouts import column
from bokeh.models import Button, ColumnDataSource
from bokeh.palettes import brewer
from bokeh.plotting import figure

N = 10
df = pd.DataFrame(np.random.randint(10, 100, size=(15, N))).add_prefix('y')
source = ColumnDataSource(df)
p = figure(x_range=(0, len(df)-1), y_range=(0, 800))
p.grid.minor_grid_line_color = '#eeeeee'

names = ["y%d" % i for i in range(N)]
p.varea_stack(stackers=names, x='index', color=brewer['Spectral'][N], legend_label=names, source=source)

p.legend.items.reverse()


def update():
    df = pd.DataFrame(np.random.randint(10, 100, size=(30, N))).add_prefix('y')    # Longer length
    source.data = ColumnDataSource.from_df(df)


button = Button(label="Update")
button.on_click(update)

curdoc().add_root(column(button, p))
```

Clicking on the update button does not change the chart.

Stack trace from the console:

```
bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:176 Uncaught Error: Size mismatch
    at Object.e.assert (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:176)
    at a.select (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:190)
    at c.set_data (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:263)
    at V.set_data (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:226)
    at V.update_data (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:226)
    at V.t (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:226)
    at V.s (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:207)
    at r.emit (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:180)
    at r.emit (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:180)
    at _setv (bokeh.min.js?v=ada5d64938a63c435e537bbd83d34dfd3b1da9bfb3ca9c7b8e947f29983f2bd491f6a157b4613e727b5739396d69da58a21a56167d248c5ff99c5f3062545d11:179)
```

**Comment by pnetherwood** [comment]
The same is true for vline_stack().

**Comment by mattpap** [comment]
This is a race condition in `Expression`'s cache, which is invalidated after data arrays of new length are requested by glyphs, resulting in data size mismatch (between `x` (`index`) and `yN`).

**Comment by danwild** [comment]
I've bumped into this one also when using `bokeh.plotting.Figure.scatter`.

My plot works fine when updating the ColumnDataSource to increase the size of the columns, but threw the size mismatch error when reducing size..

[The docs suggest](https://docs.bokeh.org/en/latest/docs/user_guide/data.html#modifying-a-columndatasource) that simply assigning a new dict to the CDS should be all is needed??

Is there something I'm missing? Is there an open issue for that cache problem or more context on what that is/does?

**Comment by pnetherwood** [comment]
All you should need to do on an update is assign a new dictionary to the
data property of the CDS. I've not had a problem with scatter. Only the
vbar/varea/vline stack charts when the size changes.

**Comment by bryevdv** [comment]
@mattpap I haven't looked at this in detail yet, do you have an idea of the scope of work to fix? I would briefly delay 2.4.1 for this if it can be resolved simply.

**Comment by mattpap** [comment]
I will give it a look tomorrow.
